### PR TITLE
[extractor/0000studio] Add support for 00:00 Studio

### PR DESCRIPTION
**Pull request by Lesmiscore**
### Before submitting a *pull request* make sure you have:
- [x] At least skimmed through [contributing guidelines](https://github.com/yt-dlp/yt-dlp/blob/master/CONTRIBUTING.md#developer-instructions) including [yt-dlp coding conventions](https://github.com/yt-dlp/yt-dlp/blob/master/CONTRIBUTING.md#yt-dlp-coding-conventions)
- [x] [Searched](https://github.com/yt-dlp/yt-dlp/search?q=is%3Apr&type=Issues) the bugtracker for similar pull requests
- [x] Checked the code with [flake8](https://pypi.python.org/pypi/flake8) and [ran relevant tests](https://github.com/yt-dlp/yt-dlp/blob/master/CONTRIBUTING.md#developer-instructions)

### In order to be accepted and merged into yt-dlp each piece of code must be in public domain or released under [Unlicense](http://unlicense.org/). Check one of the following options:
- [x] I am the original author of this code and I am willing to release it under [Unlicense](http://unlicense.org/)
- [ ] I am not the original author of this code but it is in public domain or released under [Unlicense](http://unlicense.org/) (provide reliable evidence)

### What is the purpose of your *pull request*?
- [ ] Fix or improvement to an extractor (Make sure to add/update tests)
- [x] New extractor ([Piracy websites will not be accepted](https://github.com/yt-dlp/yt-dlp/blob/master/CONTRIBUTING.md#is-the-website-primarily-used-for-piracy))
- [ ] Core bug fix/improvement
- [ ] New feature (It is strongly [recommended to open an issue first](https://github.com/yt-dlp/yt-dlp/blob/master/CONTRIBUTING.md#adding-new-feature-or-making-overarching-changes))

---

### Description of your *pull request* and other information

This adds support for 00:00 Studio. Livestream support is not implemented as it uses WebRTC to deliver video streams.

### CODE REVIEW

File: yt_dlp/extractor/fourzerostudio.py [hunk 1]
```diff
@@ -0,0 +1,111 @@
...
+            if 'mp4' in traverse_obj(m, ('formats', ..., 'ext')):
+                info = m
+                break
+        else:
+            self.report_warning('Failed to find a desired media element. Falling back to using NUXT data.')
+            info = {
+                'formats': [{
+                    'ext': 'mp4',
+                    'url': url,
+                } for url in clip_info.get('mediaFiles') if url],
```
Comment (Ashish0804) [inline yt_dlp/extractor/fourzerostudio.py:101]:
```suggestion
                } for url in clip_info.get('mediaFiles') or [] if url],
```

File: yt_dlp/extractor/fourzerostudio.py [hunk 2]
```diff
@@ -0,0 +1,111 @@
...
+
+    def _real_extract(self, url):
+        video_id, uploader_id = self._match_valid_url(url).group('id', 'uploader_id')
+        webpage = self._download_webpage(url, video_id)
+        nuxt_data = self._search_nuxt_data(webpage, video_id, full_data=True)
+
+        pcb = traverse_obj(nuxt_data, ('ssrRefs', lambda _, v: v['__typename'] == 'PublicCreatorBroadcast'), get_all=False)
+        comments = traverse_obj(nuxt_data, ('ssrRefs', ..., lambda _, v: v['__typename'] == 'PublicCreatorBroadcastComment'))
+
+        formats = self._extract_m3u8_formats(
```
Comment (Ashish0804) [inline yt_dlp/extractor/fourzerostudio.py:46]: use `_and_subtitles` even if there are no subs right now

File: yt_dlp/extractor/fourzerostudio.py [hunk 3]
```diff
@@ -0,0 +1,111 @@
...
+
+        return {
+            'id': video_id,
+            'title': pcb.get('title'),
+            'age_limit': 18 if pcb.get('isAdult') else None,
+            'timestamp': unified_timestamp(pcb.get('finishTime')),
+            'release_timestamp': unified_timestamp(pcb.get('createdAt')),
+            'thumbnails': [{
+                'url': pcb.get('thumbnailUrl'),
+                # the file seems a PNG file
```
Comment (Ashish0804) [inline yt_dlp/extractor/fourzerostudio.py:58]: no need for this comment

File: yt_dlp/extractor/fourzerostudio.py (hunk 1 above)
Comment (Ashish0804) [inline yt_dlp/extractor/fourzerostudio.py:101]: fucked formatting on GH side or is it in the file? 
The else block should be 1 tab right

File: yt_dlp/extractor/fourzerostudio.py [hunk 4]
```diff
@@ -0,0 +1,111 @@
+from .common import InfoExtractor
+from ..utils import (
+    traverse_obj,
+    unified_timestamp,
+)
+
+
+# Livestreams support cannot be implemented as this service uses WebRTC for it
+# Support request may still qualify
```
Comment (Ashish0804) [inline yt_dlp/extractor/fourzerostudio.py:9]: Not needed

File: yt_dlp/extractor/fourzerostudio.py (hunk 1 above)
Comment (Lesmiscore) [reply yt_dlp/extractor/fourzerostudio.py:101]: this else block is for fallback of the `for` loop. there are one dummy `<audio>` tag and one which this loop is looking for, so indenting by 1 will run the fallback block twice at most

**Review by Ashish0804 — APPROVED** [review]
lgtm

File: yt_dlp/extractor/common.py [hunk 5]
```diff
@@ -1568,7 +1568,7 @@ def _search_nextjs_data(self, webpage, video_id, *, transform_source=None, fatal
                 webpage, 'next.js data', fatal=fatal, **kw),
             video_id, transform_source=transform_source, fatal=fatal)
 
-    def _search_nuxt_data(self, webpage, video_id, context_name='__NUXT__'):
+    def _search_nuxt_data(self, webpage, video_id, context_name='__NUXT__', full_data=False):
```
Comment (pukkandan) [inline yt_dlp/extractor/common.py:1571]:
```suggestion
    def _search_nuxt_data(self, webpage, video_id, context_name='__NUXT__', return_full_data=False):
```

File: yt_dlp/extractor/fourzerostudio.py [hunk 6]
```diff
@@ -0,0 +1,108 @@
...
+
+
+class FourZeroStudioBaseIE(InfoExtractor):
+
+    @staticmethod
+    def _extract_uploader_info(nuxt_data):
+        uploader_info = traverse_obj(nuxt_data, ('ssrRefs', lambda _, v: v['__typename'] == 'PublicUser'), get_all=False)
+        return {
+            'uploader': uploader_info.get('username'),
+        }
```
Comment (pukkandan) [inline yt_dlp/extractor/fourzerostudio.py:15]:
```suggestion
        return {
            'uploader': traverse_obj(nuxt_data, (
                'ssrRefs', lambda _, v: v['__typename'] == 'PublicUser', 'username'
            ), get_all=False),
        }
```
The function (and the baseIE) isn't even needed imo

File: yt_dlp/extractor/fourzerostudio.py [hunk 7]
```diff
@@ -0,0 +1,108 @@
...
+    }]
+
+    def _real_extract(self, url):
+        video_id, uploader_id = self._match_valid_url(url).group('id', 'uploader_id')
+        webpage = self._download_webpage(url, video_id)
+        nuxt_data = self._search_nuxt_data(webpage, video_id, full_data=True)
+
+        pcb = traverse_obj(nuxt_data, ('ssrRefs', lambda _, v: v['__typename'] == 'PublicCreatorBroadcast'), get_all=False)
+        comments = traverse_obj(nuxt_data, ('ssrRefs', ..., lambda _, v: v['__typename'] == 'PublicCreatorBroadcastComment'))
```
Comment (pukkandan) [inline yt_dlp/extractor/fourzerostudio.py:41]: comments returned by infodict is supposed to have a specific form
```
    comments:       A list of comments, each with one or more of the following
                    properties (all but one of text or html optional):
                        * "author" - human-readable name of the comment author
                        * "author_id" - user ID of the comment author
                        * "author_thumbnail" - The thumbnail of the comment author
                        * "id" - Comment ID
                        * "html" - Comment as HTML
                        * "text" - Plain text of the comment
                        * "timestamp" - UNIX timestamp of comment
                        * "parent" - ID of the comment this one is replying to.
                                     Set to "root" to indicate that this is a
                                     comment to the original video.
                        * "like_count" - Number of positive ratings of the comment
                        * "dislike_count" - Number of negative ratings of the comment
                        * "is_favorited" - Whether the comment is marked as
                                           favorite by the video uploader
                        * "author_is_uploader" - Whether the comment is made by
                                                 the video uploader
```

File: yt_dlp/extractor/fourzerostudio.py [hunk 8]
```diff
@@ -0,0 +1,108 @@
...
+
+        formats, subs = self._extract_m3u8_formats_and_subtitles(
+            pcb.get('archiveUrl'), video_id, ext='mp4')
```
Comment (pukkandan) [inline yt_dlp/extractor/fourzerostudio.py:44]:
```suggestion
        formats, subs = self._extract_m3u8_formats_and_subtitles(pcb['archiveUrl'], video_id, ext='mp4')
```

File: yt_dlp/extractor/fourzerostudio.py [hunk 9]
```diff
@@ -0,0 +1,108 @@
...
+        return {
+            'id': video_id,
+            'title': pcb.get('title'),
+            'age_limit': 18 if pcb.get('isAdult') else None,
+            'timestamp': unified_timestamp(pcb.get('finishTime')),
+            'release_timestamp': unified_timestamp(pcb.get('createdAt')),
+            'thumbnails': [{
+                'url': pcb.get('thumbnailUrl'),
+                'ext': 'png',
+            }],
```
Comment (pukkandan) [inline yt_dlp/extractor/fourzerostudio.py:56]:
```suggestion
            'thumbnails': [{
                'url': pcb['thumbnailUrl'],
                'ext': 'png',
            }] if pcb.get('thumbnailUrl') else None,
```

File: yt_dlp/extractor/fourzerostudio.py [hunk 10]
```diff
@@ -0,0 +1,108 @@
...
+        webpage = self._download_webpage(url, video_id)
+        nuxt_data = self._search_nuxt_data(webpage, video_id, full_data=True)
+
+        clip_info = traverse_obj(nuxt_data, ('ssrRefs', lambda _, v: v['__typename'] == 'PublicCreatorArchivedClip'), get_all=False)
+
+        for m in self._parse_html5_media_entries(url, webpage, video_id):
+            if 'mp4' in traverse_obj(m, ('formats', ..., 'ext')):
+                info = m
+                break
+        else:
```
Comment (pukkandan) [inline yt_dlp/extractor/fourzerostudio.py:92]:
```suggestion
        info = next((
            m for m in self._parse_html5_media_entries(url, webpage, video_id)
            if 'mp4' in traverse_obj(m, ('formats', ..., 'ext'))
        ), None)
        if not info:
```

File: yt_dlp/extractor/fourzerostudio.py [hunk 11]
```diff
@@ -0,0 +1,108 @@
...
+            }
+        info.update({
+            'id': video_id,
+            'title': clip_info.get('clipComment'),
+            'timestamp': unified_timestamp(clip_info.get('createdAt')),
+            'like_count': clip_info.get('likeCount'),
+            'uploader_id': uploader_id,
+            **self._extract_uploader_info(nuxt_data),
+        })
+        return info
```
Comment (pukkandan) [inline yt_dlp/extractor/fourzerostudio.py:108]:
```suggestion
        return {
            **info,
            'id': video_id,
            'title': clip_info.get('clipComment'),
            'timestamp': unified_timestamp(clip_info.get('createdAt')),
            'like_count': clip_info.get('likeCount'),
            'uploader_id': uploader_id,
            **self._extract_uploader_info(nuxt_data),
        }
```

**Review by pukkandan — CHANGES_REQUESTED** [review]

**Pull request merged**